File: Go/island-perimeter.go
func islandPerimeter(grid [][]int) int {
    rows, cols := len(grid), len(grid[0])

    perimeter := 0
    for i := 0; i <= rows; i++ {
        left := 0
        for j := 0; j <= cols; j++ {
            curr, up := 0, 0
            if j < cols {
                if i > 0 {
                    up = grid[i - 1][j]
                }
                if i < rows {
                    curr = grid[i][j]
                }
            }
            if curr != up {
                perimeter++
            }
            if curr != left {
                perimeter++
            }
            left = curr
        }
    }
    return perimeter
}

File: C++/island-perimeter.cpp
class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();
        int perimeter = 0;
        for (int r = 0; r <= rows; r++) {
            int left = 0;
            for (int c = 0; c <= cols; c++) {
                int curr = (c < cols && r < rows) ? grid[r][c] : 0;
                int up = (c < cols && r > 0) ? grid[r - 1][c] : 0;
                perimeter += (curr != up) + (curr != left);
                left = curr;
            }
        }
        return perimeter;
    }
};